src/rust/src/pkcs7.rs

```diff
@@ -407,6 +566,39 @@ fn smime_canonicalize(data: &[u8], text_mode: bool) -> (Cow<'_, [u8]>, Cow<'_, [
     }
 }
 
+fn smime_decanonicalize(data: &[u8], text_mode: bool) -> Cow<'_, [u8]> {
+    let mut new_data = vec![];
+    let mut last_idx = 0;
+
+    // Remove the header if text_mode is true
+    let data = if text_mode {
+        let header = b"Content-Type: text/plain\r\n\r\n";
+        if data.starts_with(header) {
+            &data[header.len()..]
+        } else {
```

Comment:
If the header isn't present in text mode, should that be an error?

Reply:
I have no idea how OpenSSL handles it for now. I'll check it out when I have time and let you know.

Reply:
After some checking with my version of OpenSSL (on Windows!), it seems the behavior of the `-text` and `-binary` options are much weirder than what was expected on `decrypt.` They are not symmetrical to their usage in `encrypt` :
- No options passed : replaces the \n into \r\n
- `-binary` : leaves the \n as such, do not replace them by \r\n
- `-text` :  Removes the header, raises an error if there is none (your comment is correct)
- Both options can be combined, where in `encrypt`, Binary was taking precedence.

This implies that if you encrypt then decrypt data using `openssl smime` and the `text` option, you can start with `Hello, world!\n` and end up with `Hello, world!\r\r\n` 😢 

I was mostly using `-binary` optons in my use cases, so I never quite found out. Adding this right now.

Reply:
@nitneuqr this looks similar to a bug I encountered while implementing s/mime signing: https://github.com/openssl/openssl/issues/23886. There is also a comment on that issue from someone finding the same bug during encryption

Reply:
Thanks for that. To me, the documentation for the `-binary` option makes sense, at least during the `-encrypt` part. But for decryption, it doesn't. The comment exactly describes my issue 😄

> -binary
> 
> Normally the input message is converted to "canonical" format which is effectively using CR and LF as end of line: as required by the S/MIME specification. When this option is present no translation occurs. This is useful when handling binary data which may not be in MIME format.

To be honest, for this PR I'd stick with the current behavior of OpenSSL (even if seems incorrect). What's your opinion on this?

Reply:
> To be honest, for this PR I'd stick with the current behavior of OpenSSL (even if seems incorrect). What's your opinion on this?

I think it's up to Alex and Paul. For context, the decision for signing [was to follow the RFC](https://github.com/pyca/cryptography/issues/10501#issuecomment-2212032777), rather than imitate what OpenSSL does.

Reply:
I think we're confusing the format of the output file (which is what is supposed to control `-crlfeol` in OpenSSL) with the format of the ciphered message inside it (which is controlled by `-text` or `-binary`), at least for encryption/decryption.

Right now, the `-crlfeol` parameter does not seem to have any effect in decryption, so I'm ignoring it for now.